### PR TITLE
Add CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by manuelbuil**
Add a contribution guide

### CODE REVIEW

File: CONTRIBUTING.md
Comment (thomasferrandiz) [inline CONTRIBUTING.md:11]: I think we should mention that generally `go generate` and ` go mod tidy` outputs should be in dedicated commits

Comment (manuelbuil) [reply CONTRIBUTING.md:11]: Good point Thomas,thanks! Let me add that

File: CONTRIBUTING.md
Comment (superseb) [inline CONTRIBUTING.md:31]: This is not enough context for a contributor. `A commit` can be core RKE code that applies to everything, which could be made k8s version (minor/patch) specific or only applies to a subset like an add-on which is already k8s version specific. This needs to be split up in the bigger currently known changes we have and address those.

Comment (manuelbuil) [reply CONTRIBUTING.md:31]: Good point but I don't get what you mean by `needs to be split up in the bigger currently known changes we have and address those`. I'll try to rewrite it to cover the different backport cases

Comment (superseb) [reply CONTRIBUTING.md:31]: We know what kind of changes we have, these need to be described in a bit more detail so the reader can identify what change falls under what category. If the reader can do that, the reader can also identify where it needs to be backported.

We have RKE code backports, RKE k8s version backports and add-on templates backports at minimum which all have their own "policy" or at least it needs to be described.

Comment (manuelbuil) [reply CONTRIBUTING.md:31]: This is what I have written.
RKE code backports:
```
By design, RKE code supports all active Kubernetes minor versions, so the commit in the RKE project is automatically available for all. You just need to remember to contribute the commit in all the active RKE releases which are developed in `release vX.Y` branches
```
add-on templates backports:
```
if the commit is done in a dependent project, such as an add-on, the `data.json` file will need to be updated accordingly, so that all Kubernetes minor versions supported by RKE consume the add-on version that includes the commit
```
Could you elaborate more about the `RKE k8s version backports` case? I am probably missing that one but I'm not sure where is the commit in that case. Thanks Seb!

Comment (superseb) [reply CONTRIBUTING.md:31]: I think this section needs to be divided in categories of changes, that we can identify and refer to in a way that is clear enough that we all know what is what. Sentences like `A commit should be available in all active Kubernetes minor versions` do not clarify enough on what the goal is.

This needs a bit of thinking to get everything cleared up, I've started a bit below:

- RKE code change is, unless scoped in the code to specific Kubernetes versions, available generically. For example, better/more/less logging statements can be merged and will be available to every invocation of RKE. Kubernetes versions are not related in this way because new RKE code gets executed wether you have a current-2, current-1 or current patch and or minor version). Backporting this means copying the code from `release/v1.X` to `release/v1.X-1`, and will be available to every invocation of the new RKE code for `release/v1.X-1`.
- RKE code changes that are scoped to Kubernetes versions/Kubernetes component versions. See https://github.com/rancher/rke/blob/645e80840596a8616c2a1f35cea0c655ac700fa9/cluster/plan.go#L60 as an example. This is outside of KDM.

Both changes above should be backported if they are related to versions that are supported in both `release/v1.X` and `release/v1.X-1` because we don't want users to have differences between provisioning a cluster in `release/v1.X-1` or `release/v1.X`. (If a flag is changed between versions that is only available in one version, an edit to the cluster will trigger provisioning (makes sense) but it will then also replace the component with the newly added flag to be removed if its not in `release/v1.X` (or vice versa)

This is just a beginning, and we don't need to flesh it out completely but this is the basics that we need to nail down to have a starting point. I am also fine with starting with what we have currently but I don't think its a good enough starting point, I'd rather take it a step further and start with more than this.

I also prefer having more clear statements than `By active we mean minor versions that Kubernetes upstream is actively working on, which normally means 3 kubernetes minors.`. We should make a statement now and say that we do latest supported - 2 for example. That is already way more clear than the previous statement. We can fine tune from there but we need some more clarity on these topics.

Lastly, the way things are described need to be more clear. Sentence like `However, if the commit is done in a dependent project, such as an add-on, the `data.json` file will need to be updated accordingly,` is weird because we don't update `data.json`. We updated source files in KDM and generate the `data.json`. I know this sounds nit picky, but for me its the difference between having a lot of confusion between multiple readers and having clarity (saving time because its more clear)

Hope it makes sense and let me know if you get what I mean or not, we can also have a session to go over it.

Comment (manuelbuil) [reply CONTRIBUTING.md:31]: Thanks for the explanation. My intention was not to cover all the different options in detail as then the document will probably become too long. My experience tells me that too long documents have the danger that nobody ever reads them, thus making this PR a bit useless. But let me try to cover most or all backport use cases and see what we get

Comment (manuelbuil) [reply CONTRIBUTING.md:31]: I made an update going more into detail and with example. The result is not as long as I anticipated :)

**Comment by thomasferrandiz** [comment]
Should we also mention opening an issue in `rancher/rancher` for tracking?

**Comment by manuelbuil** [comment]
> Should we also mention opening an issue in `rancher/rancher` for tracking?

Yes!

File: CONTRIBUTING.md
Comment (a-blender) [inline CONTRIBUTING.md]: 1. Can we create a PR template for rancher/rke? It'd be a lot easier to direct contributors on what/what is required to add to a PR.
2. As I understand it, we are focusing much more on `what code has been modified/added` in the PR description to be able to catch regressions quicker, not just adding repro steps to the issue. I'd add a splurge about that to make that obvious to contributors. Our focus on regression prevention needs to extend to community PRs too.

File: CONTRIBUTING.md
Comment (a-blender) [inline CONTRIBUTING.md:23]: Are the two first points backwards? If a PR has a single commit with added/updated logic `Create a merge commit` and if 1+ logic commits then `Rebase and merge`. Will also wait for @superseb's comments to be addressed and re-review this section after that.

File: CONTRIBUTING.md
Comment (a-blender) [inline CONTRIBUTING.md]: `release/v1.3` is not being actively maintained but will have critical bugs backported to it and released again if it's necessary to do so. It's not EOL.

File: CONTRIBUTING.md
Comment (a-blender) [inline CONTRIBUTING.md]: Awk wording. Suggest: _Each track releases an RKE binary._

File: CONTRIBUTING.md
Comment (a-blender) [inline CONTRIBUTING.md]: Suggest: _If a Kubernetes version is **supported**, that means we can deploy a cluster with that Kubernetes version using the binary included in that RKE release._

File: CONTRIBUTING.md
Comment (a-blender) [inline CONTRIBUTING.md]: I'd add a link to the code that checks the version

**Review by a-blender — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md]: re: 1, yes, we could do that but in another PR. Could you create an issue about it?

re: 2. if you are referring to the line `If it is a bug fix, please describe how to reproduce it in the issue`. The reason is mainly that we should triage the issue to verify that it is indeed a general RKE1 issue. A lot of users have problems which end up being a misconfiguration in their hosts

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md:23]: Good catch! I am missing the part where the commits should be squashed together once the PR has been approved

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md]: Thanks for this information, I was not aware of it. When does a version become EOL?

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md]: You mean github.com/coreos/go-semver/semver? We are using that type and call functions such as https://github.com/rancher/rke/blob/release/v1.5/cluster/validation.go#L752

File: CONTRIBUTING.md
Comment (a-blender) [reply CONTRIBUTING.md]: 1. Yes! https://github.com/rancher/rke/issues/3442
2. True, but that's an addition to the issue and this section is on 'Opening PRs and organizing commits' so I think it'd be helpful to add docs that will help the contributor focus on _what code/additions were added_ in the PR itself.

File: CONTRIBUTING.md
Comment (a-blender) [reply CONTRIBUTING.md]: Rancher uses rke under the hood so it's tied to Rancher and the EOL for supported Kubernetes versions. For example, the `release/v1.3` line corresponds with Rancher 2.6 clusters and according to the support matrix, https://www.suse.com/suse-rancher/support-matrix/all-supported-versions/rancher-v2-6-13/ the latest Rancher 2.6 only supports up to k8s 1.24 clusters.

We are far beyond that now with k8s 1.26 reaching active EOL (and the maintenance phase) on Feb 28, 2024 https://kubernetes.io/releases/. Customers using rke 1.3.x are most likely in a position to upgrade than ask for new features or enhancements on their current version of rke. That being said, we still need to support customers on old versions so they still require passive maintenance.

cc @kinarashah do we have any rke versions that are post-maintenance EOL?

File: CONTRIBUTING.md
Comment (a-blender) [reply CONTRIBUTING.md]: Or this https://github.com/rancher/rke/blob/c56fb8c42a079d569dd402c1243a5ae16b4004c9/cluster/validation.go#L748 but yes

File: CONTRIBUTING.md
Comment (snasovich) [inline CONTRIBUTING.md]: > To keep things simple, if you're fixing a small issue, you can simply submit a PR and we will pick it up.
2 things:
1. I would still prefer to have an issue to track even "small issues". The only exceptions would be things like typos / misc changes. Anything that fixes an actual issue or implements a feature should an issue IMO.
2. I would remove "we will pick it up" as it establishes commitment to review any submitted PR.

File: CONTRIBUTING.md
Comment (snasovich) [inline CONTRIBUTING.md]: Minor wording changes, but I agree with @a-blender 's suggestion to establish PR template.

```suggestion
Each PR needs to be linked to a Github issue in [rancher project](https://github.com/rancher/rancher) which provides as much detail as possible on the change and its motivation. If it is a bug fix, please describe how to reproduce it in the issue. Make the Team or Project Manager aware of the GitHub issue so that they can set the correct milestone. Milestones are very important for the release and QA tracking.
```

File: CONTRIBUTING.md
Comment (snasovich) [inline CONTRIBUTING.md]: I don't think we generally following this approach and using PR description in lieu of "detailed body". Our commit messages are generally short - for better or worse.

File: CONTRIBUTING.md
Comment (snasovich) [inline CONTRIBUTING.md]:
```suggestion
When addressing review feedback, it is helpful to the reviewer if additional changes are made in new commits. This allows the reviewer to easily see the delta between what they previously reviewed and the changes you added to address their feedback. Once positively reviewed, those commits should be squashed into one.
```

File: CONTRIBUTING.md
Comment (snasovich) [inline CONTRIBUTING.md:24]: Suggesting to specifically call out the case of `go generate` commits here so they are not accidentally squashed.

File: CONTRIBUTING.md
Comment (snasovich) [inline CONTRIBUTING.md]: Suggesting to use relative link so we're not pointing to specific branch (default branch is already `release/v1.5` while this URL points to `release/v1.4`).

File: CONTRIBUTING.md
Comment (snasovich) [inline CONTRIBUTING.md]: Minor typos
```suggestion
To understand how backports should be pulled through, RKE project releases and what is included in each release should be described. RKE in general maintains two tracks in parallel which are differentiated by the minor version. Each of these tracks is developed in two different GitHub branches:
```

**Review by snasovich — COMMENTED** [review]
Thank you, @manuelbuil . Posted some comments. Not using "Request changes" as I don't want my re-review to block merging.

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md]: My understanding was that once k8s upstream version reaches EOL, we don't do any support to our customers or maybe just best effort. This will change when we announce the LTS kubernetes product which I don't know when will it happen

Comment (manuelbuil) [reply CONTRIBUTING.md:68]: My understanding was that once k8s upstream version reaches EOL, we don't do any support to our customers or maybe just best effort. This will change when we announce the LTS kubernetes product which I don't know when will it happen

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md]: 1. Yes, let's change that. The idea was to facilitate users contributing. Having to create an issue is another step which could be perceived as a burden. But as we don't expect RKE to become a "popular" open source project like k3s, we can remove that
2. I believe we should commit to review any submitted PR. Why not?

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md]: ok, I can change that. I don't care where things are described as long as they are described :)

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md:24]: good point

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md]: What do you mean by relative link? I thought I always need to specify the release in the url to access the file (README.md in this case)

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md]: I change opinion :). I think it is best to add the PR Template to this PR because it is related

Comment (manuelbuil) [reply CONTRIBUTING.md]: PR template proposal now included in this PR

File: CONTRIBUTING.md
Comment (kinarashah) [inline CONTRIBUTING.md]: Thought we also need to move the issue to `To-Test`, is that no longer applicable?

File: CONTRIBUTING.md
Comment (kinarashah) [inline CONTRIBUTING.md]: nit typo: `these tracks s developped`

File: CONTRIBUTING.md
Comment (kinarashah) [inline CONTRIBUTING.md:52]: Sometimes there isn't a reason to backport it/ Sometimes there is a reason not to backport it

File: CONTRIBUTING.md
Comment (kinarashah) [inline CONTRIBUTING.md]: Are we now creating `release/v1.x` milestone under rancher/rancher? We usually create this milestone under rke.

File: CONTRIBUTING.md
Comment (kinarashah) [inline CONTRIBUTING.md]: > In this case, as the same code is executed for all Kubernetes versions, unless scoped, there is no need to create extra PRs to effectively backport the code to all Kubernetes versions. 

This line needs to be modified, backporting for rke1 is decided based on rke1 versions and once backported - it'll apply to k8s versions it supports. Branches aren't based on k8s versions.

File: CONTRIBUTING.md
Comment (kinarashah) [reply CONTRIBUTING.md]: Could you link this to templates.go file that maps the versions to k8s versions instead?

**Review by kinarashah — COMMENTED** [review]
This is a good place to start @manuelbuil, and we can elaborate on the addon section in another PR.

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md]: Ah good point! It would be [zube] To Test, right?

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md]: If I understood correctly:
* milestones are created in rancher/rancher
* branch `releases/v1.x` is created in rancher/rke project
And they are connected. For example:
* milestone [v2.8-KDM-Nov-2023-patches](https://github.com/rancher/rancher/issues?q=is%3Aopen+is%3Aissue+milestone%3Av2.8-KDM-Nov-2023-patches) in rancher/rancher maps to branch [release/v1.5](https://github.com/rancher/rke/tree/release/v1.5) in rancher/rke
* milestone [v2.7-KDM-Nov-2023-patches](https://github.com/rancher/rancher/issues?q=is%3Aopen+is%3Aissue+milestone%3Av2.7-KDM-Nov-2023-patches) in rancher/rancher maps to branch [release/v1.4](https://github.com/rancher/rke/tree/release/v1.4) in rancher/rke
right?

Comment (manuelbuil) [reply CONTRIBUTING.md]: Assuming what I wrote is correct, let me add one extra sentence to explain this. I can see how it could be confusing for the reader

File: CONTRIBUTING.md
Comment (manuelbuil) [reply CONTRIBUTING.md]: I was trying to explain that there is no need to backport stuff to different k8s versions (as we do in rke2/k3s) because the PR will apply to k8s versions supported by the rke1 version. Let me rephrase it a bit better

**Review by briandowns — APPROVED** [review]

File: CONTRIBUTING.md
Comment (jiaqiluo) [inline CONTRIBUTING.md:8]: AFAIK, there are multiple formats of the milestones. I suggest removing the last sentence.

File: CONTRIBUTING.md
Comment (jiaqiluo) [inline CONTRIBUTING.md:29]: The label "[zube] To Test" is deprecated. Now we change the status of the issue to "Ready for QA" (or similar state) in the Project(s) that the issue is associated with.  GitHub UI shows the Projects section under the Labels section on the right of the page.

File: CONTRIBUTING.md
Comment (jiaqiluo) [inline CONTRIBUTING.md:35]: Could you rephrase the first sentence which is confusing to read (at leat to me)?

File: CONTRIBUTING.md
Comment (jiaqiluo) [inline CONTRIBUTING.md:52]:
```suggestion
When doing a commit in RKE code, we must carefully think if we want that commit in both tracks and hence backported. Sometimes, there is a reason not to backport it, for example, a flag that we would like to limit to only the newer track. However, that is exceptional and normally commits end up in both branches as a way to reduce the differences between those and simplify RKE maintainability. Therefore, usually we backport a commit between both RKE tracks. In that case, the following steps should be applied:
```

**Review by jiaqiluo — APPROVED** [review]
Overall LGTM, and I think this is a good addition to the project! 
I also left a few comments.

**Review by jiaqiluo — CHANGES_REQUESTED** [review]
request changes

**Pull request closed without merging**
